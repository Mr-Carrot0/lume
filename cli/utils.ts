import lume from "../mod.ts";
import { log } from "../core/utils/log.ts";
import Site from "../core/site.ts";

/** Create a site instance */
export async function createSite(_config?: URL): Promise<Site> {
  if (!_config) {
    return lume();
  }
  log.info(`Loading config file <gray>${_config}</gray>`);
  const mod = await import(_config.toString());
  const site = mod.default as Site | undefined;
  if (!(site instanceof Site)) {
    log.fatal(
      `[Lume] Missing Site instance! Ensure your config file does export the Site instance as default.`,
    );
    throw new Error("Site instance is not found");
  }
  return site;
}

/** Create a site intance and build it */
export async function buildSite(site: Site): Promise<void> {
  performance.mark("start");
  await site.build();
  performance.mark("end");

  log.info(`üçæ Site built into <gray>${site.options.dest}</gray>`);
  const duration = performance.measure("duration", "start", "end").duration /
    1000;
  const total = site.pages.length + site.files.length;
  log.info(
    `  <gray>${total} files generated in ${duration.toFixed(2)} seconds</gray>`,
  );
}
