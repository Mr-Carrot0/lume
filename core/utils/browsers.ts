/**
 * Code generated from:
 *
 * ```ts
 * import { getCompatibleVersions } from "npm:baseline-browser-mapping@2.4.4";
 *
 * const results = getCompatibleVersions({
 *   widelyAvailableOnDate: "2025-05-07", // Lume 3.0.0 release date
 * });
 * ```
 *
 * Not run directly to avoid importing this package that requires read access.
 */
export const browsers = {
  chrome: [107, 0],
  chrome_android: [107, 0],
  edge: [107, 0],
  firefox: [104, 0],
  firefox_android: [104, 0],
  safari: [16, 0],
  safari_ios: [16, 0],
} as Record<string, [number, number]>;

export function version(browser: [number, number]): number {
  return (browser[0] << 16) | (browser[1] << 8);
}
export function versionString(browser: [number, number]): string {
  return `${browser[0]}.${browser[1]}`;
}
