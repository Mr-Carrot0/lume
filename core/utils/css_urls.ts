type UrlWalker = (url: string, type: string) => string | Promise<string>;
type Position = [string, string, number, number];
type Index = [string, number];

export async function walkUrls(
  code: string,
  walker: UrlWalker,
): Promise<string> {
  let result = "";
  let index = 0;
  let position = next(code, index);

  while (position) {
    const [type, url, start, end] = position;
    const newUrl = await walker(url, type);
    const quotedUrl = url.startsWith("data:")
      ? `'${newUrl.replaceAll("'", "\\'")}'`
      : `"${newUrl.replaceAll('"', '\\"')}"`;
    result += `${code.slice(index, start)}${quotedUrl}`;
    index = end;
    position = next(code, index);
  }

  return result + code.slice(index);
}

function next(code: string, index = 0): Position | undefined {
  const nextUrl = code.indexOf("url(", index);
  const nextComment = code.indexOf("/*", index);
  const nextImport = code.indexOf("@import ", index);

  let found: Index | undefined;

  if (nextUrl !== -1) {
    found = ["url", nextUrl];
  }
  if (nextComment !== -1 && (!found || nextComment < found[1])) {
    found = ["comment", nextComment];
  }
  if (nextImport !== -1 && (!found || nextImport < found[1])) {
    found = ["import", nextImport];
  }

  if (!found) {
    return;
  }

  // If it's a comment, find the end of it
  if (found[0] === "comment") {
    const end = code.indexOf("*/", found[1]);
    if (end === -1) {
      return;
    }
    return next(code, end + 2);
  }

  if (found[0] === "import") {
    const double = code.indexOf('"', found[1]);
    const single = code.indexOf("'", found[1]);
    const urlFn = code.indexOf("url(", found[1]);

    let foundImport: Index | undefined;

    if (double !== -1) {
      foundImport = ["double", double];
    }
    if (single !== -1 && (!foundImport || single < foundImport[1])) {
      foundImport = ["single", single];
    }

    // It's has a url(), so delegate to the url() logic
    if (urlFn !== -1 && (!foundImport || urlFn < foundImport[1])) {
      return next(code, urlFn);
    }

    if (!foundImport) {
      return;
    }

    const start = foundImport[1];
    const end =
      code.indexOf(foundImport[0] === "double" ? '"' : "'", start + 1) + 1;
    return ["url", code.slice(start + 1, end - 1), start, end];
  }

  if (found[0] === "url") {
    const start = code.indexOf("(", found[1]) + 1;
    let url = code.slice(start).trim();
    const quote = url.startsWith('"') || url.startsWith("'")
      ? url[0]
      : undefined;
    url = quote ? url.slice(1) : url;
    let end = findUrlEnd(url, quote);
    url = url.slice(0, end);
    end += start + (quote ? 2 : 0);
    url = url.replace(/\\'/g, "'").replace(/\\"/g, '"');
    return ["url", url, start, end];
  }
}

function findUrlEnd(code: string, quote?: string): number {
  if (!quote) {
    return code.indexOf(")");
  }

  let end = code.indexOf(quote);

  while (end !== -1 && code[end - 1] === "\\") {
    end = code.indexOf(quote, end + 1);
  }

  return end === -1 ? code.indexOf(")") : end;
}
