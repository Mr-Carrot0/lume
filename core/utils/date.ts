/**
 * Returns the date of the git commit that created or modified the file.
 * Thanks to https://github.com/11ty/eleventy/blob/8dd2a1012de92c5ee1eab7c37e6bf1b36183927e/src/Util/DateGitLastUpdated.js
 */
export function getGitDate(
  type: "created" | "modified",
  file: string,
): Date | undefined {
  const args = type === "created"
    ? ["log", "--diff-filter=A", "--follow", "-1", "--format=%at", "--", file]
    : ["log", "-1", "--format=%at", "--", file];

  const { stdout, success } = new Deno.Command("git", { args }).outputSync();

  if (!success) {
    return;
  }
  const str = new TextDecoder().decode(stdout);

  if (str) {
    return parseDate(parseInt(str) * 1000);
  }
}

/** Parse a string or number (of miliseconds) to UTC Date */
export function parseDate(date: string | number): Date {
  return new Date(getZonedDateTime(date).epochMilliseconds);
}

/** Parse a string or number (of miliseconds) to a zoned datetime */
export function getZonedDateTime(
  date: string | number,
  timezone = "UTC",
): Temporal.ZonedDateTime {
  if (typeof date === "number") {
    return Temporal.Instant.fromEpochMilliseconds(date).toZonedDateTimeISO(
      timezone,
    );
  }

  try {
    return Temporal.Instant.from(date).toZonedDateTimeISO(timezone);
  } catch {
    return Temporal.PlainDateTime.from(date).toZonedDateTime(timezone);
  }
}
