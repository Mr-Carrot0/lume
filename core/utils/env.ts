const envVars = new Map<string, string>();

export function setEnv(name: string, value: string) {
  if (allowedEnvVars()) {
    Deno.env.set(name, value);
  }
  envVars.set(name, value);
}

export function env<T>(name: string): T | undefined {
  const value = allowedEnvVars()
    ? envVars.get(name) ?? Deno.env.get(name)
    : envVars.get(name);

  if (typeof value === "undefined") {
    return undefined;
  }

  switch (value.toLowerCase()) {
    case "true":
    case "on":
    case "1":
      return true as T;

    case "false":
    case "off":
    case "0":
      return false as T;

    default:
      return value as T;
  }
}

let allowed: boolean | undefined;
function allowedEnvVars(): boolean {
  if (allowed === undefined) {
    allowed = Deno.permissions.querySync?.({ name: "env" }).state === "granted";
  }
  return allowed;
}
