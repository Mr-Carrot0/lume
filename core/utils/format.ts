const byteUnit = Intl.NumberFormat("en", {
  notation: "compact",
  style: "unit",
  unit: "byte",
  unitDisplay: "narrow",
});

export function bytes(bytes: number) {
  return byteUnit.format(bytes);
}

const percentageUnit = Intl.NumberFormat("en", {
  style: "percent",
  minimumFractionDigits: 1,
  maximumFractionDigits: 1,
});

export function percentage(initialValue: number, finalValue: number) {
  if (initialValue === finalValue) {
    return "0%";
  }

  const ratio = -(1 - finalValue / initialValue);

  if (ratio < 0) {
    return percentageUnit.format(ratio);
  }

  return "+" + percentageUnit.format(ratio);
}

const secondUnit = Intl.NumberFormat("en", {
  notation: "compact",
  style: "unit",
  unit: "second",
  unitDisplay: "short",
  minimumFractionDigits: 1,
  maximumFractionDigits: 1,
});

const milisecondUnit = Intl.NumberFormat("en", {
  notation: "compact",
  style: "unit",
  unit: "millisecond",
  unitDisplay: "short",
  maximumFractionDigits: 0,
});

export function duration(miliseconds: number) {
  if (miliseconds < 1000) {
    return milisecondUnit.format(miliseconds);
  }

  return secondUnit.format(miliseconds / 1000);
}
