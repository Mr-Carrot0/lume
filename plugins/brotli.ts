import { merge } from "../core/utils/object.ts";
import { Page } from "../core/file.ts";
import { compress } from "../deps/brotli.ts";

import type { Extensions } from "../core/utils/path.ts";
import type Site from "../core/site.ts";

export interface Options {
  /** File extensions to compress */
  extensions?: Extensions;

  /**
   * Quality param between 0 and 11 (11 is the smallest but takes the longest to encode)
   */
  quality?: number;
}

// Default options
export const defaults: Options = {
  extensions: [".html", ".css", ".js", ".mjs", ".svg", ".json", ".xml", ".txt"],
  quality: 6,
};

/**
 * A plugin to compress files with brotli
 */
export function brotli(userOptions?: Options) {
  const options = merge(defaults, userOptions);

  return (site: Site) => {
    site.process(options.extensions, function processBrotli(pages, allPages) {
      for (const page of pages) {
        const compressedContent = compress(
          page.bytes,
          undefined,
          options.quality,
        );

        allPages.push(Page.create({
          url: page.outputPath + ".br",
          content: compressedContent,
        }));
      }
    });
  };
}

export default brotli;
