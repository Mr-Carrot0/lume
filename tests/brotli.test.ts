import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import brotli from "../plugins/brotli.ts";
import extractDate from "../plugins/extract_date.ts";

Deno.test("brotli plugin", async (t) => {
  const site = getSite({
    src: "normal",
  });

  site.add([".png", ".css", ".json"]);

  site.use(brotli());
  site.use(extractDate());

  await build(site);
  await assertSiteSnapshot(t, site);
});

Deno.test("brotli plugin with options", async (t) => {
  const site = getSite({
    src: "normal",
  });

  site.add([".png", ".css", ".json"]);

  site.use(brotli({
    extensions: [".css"],
    quality: 1,
  }));
  site.use(extractDate());

  await build(site);
  await assertSiteSnapshot(t, site);
});
