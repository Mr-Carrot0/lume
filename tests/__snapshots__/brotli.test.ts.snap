export const snapshot = {};

snapshot[`brotli plugin 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      isPage: true,
      loader: [AsyncFunction: toml],
    },
    {
      engines: 1,
      ext: ".page.ts",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".page.js",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      engines: 0,
      ext: ".page.json",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".vto",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      ext: ".png",
    },
    {
      ext: ".css",
    },
    {
      ext: ".html",
    },
    {
      ext: ".mjs",
    },
    {
      ext: ".svg",
    },
    {
      ext: ".xml",
    },
    {
      ext: ".txt",
    },
  ],
  src: [
    "/",
    "/404.md",
    "/_data.yml",
    "/favicon.png",
    "/images",
    "/images/avatar.jpg",
    "/page5.yaml",
    "/pages",
    "/pages/2020-06-21_page2.page.json",
    "/pages/2021-01-02-18-32_page4.page.ts",
    "/pages/_data",
    "/pages/_data.yml",
    "/pages/_data/colors.yml",
    "/pages/_data/documents.ts",
    "/pages/_data/drinks.js",
    "/pages/_data/names.json",
    "/pages/ghost",
    "/pages/ghost/2021-12-29-page6.md",
    "/pages/ghost/_data.yml",
    "/pages/page1.md",
    "/pages/page3.page.js",
    "/pages/subpage",
    "/pages/subpage/_data.yml",
    "/pages/subpage/page7.page.js",
    "/static.yml",
    "/styles.css",
  ],
}
`;

snapshot[`brotli plugin 2`] = `
[
  {
    entry: "/favicon.png",
    flags: [],
    outputPath: "/favicon.png",
  },
]
`;

snapshot[`brotli plugin 3`] = `
[
  {
    content: "<p>This page is exported to <code>/404.html</code>, not <code>/404/index.html</code></p>
",
    data: {
      basename: "404",
      children: "<p>This page is exported to <code>/404.html</code>, not <code>/404/index.html</code></p>
",
      content: "This page is exported to \`/404.html\`, not \`/404/index.html\`
",
      date: [],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Default site name",
      tags: "Array(0)",
      url: "/404.html",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/404",
      remote: undefined,
    },
  },
  {
    content: "Uint8Array(72)",
    data: {
      basename: "404",
      content: "Uint8Array(72)",
      page: [
        "src",
        "data",
        "asset",
      ],
      url: "/404.html.br",
    },
    src: {
      asset: true,
      ext: "",
      path: "",
      remote: undefined,
    },
  },
  {
    content: "Content of Page 2",
    data: {
      basename: "page2",
      children: "Content of Page 2",
      colors: "Array(3)",
      content: "Content of Page 2",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
        "image",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Folder overrided site name",
      tags: "Array(1)",
      title: "Page 2",
      url: "/overrided-page2/",
    },
    src: {
      asset: false,
      ext: ".page.json",
      path: "/pages/2020-06-21_page2",
      remote: undefined,
    },
  },
  {
    content: "Uint8Array(20)",
    data: {
      basename: "index",
      content: "Uint8Array(20)",
      page: [
        "src",
        "data",
        "asset",
      ],
      url: "/overrided-page2/index.html.br",
    },
    src: {
      asset: true,
      ext: "",
      path: "",
      remote: undefined,
    },
  },
  {
    content: "Content of Page 3",
    data: {
      basename: "page3",
      colors: "Array(3)",
      content: "Content of Page 3",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Folder overrided site name",
      tags: "Array(1)",
      title: "Page 3",
      url: "/page_3",
    },
    src: {
      asset: false,
      ext: ".page.js",
      path: "/pages/page3",
      remote: undefined,
    },
  },
  {
    content: "Content of Page 5",
    data: {
      basename: "page5",
      children: "Content of Page 5",
      content: "Content of Page 5",
      date: [],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Default site name",
      tags: "Array(0)",
      title: "Page 5",
      url: "/page5/",
    },
    src: {
      asset: false,
      ext: ".yaml",
      path: "/page5",
      remote: undefined,
    },
  },
  {
    content: "Uint8Array(20)",
    data: {
      basename: "index",
      content: "Uint8Array(20)",
      page: [
        "src",
        "data",
        "asset",
      ],
      url: "/page5/index.html.br",
    },
    src: {
      asset: true,
      ext: "",
      path: "",
      remote: undefined,
    },
  },
  {
    content: "Content of Page 7",
    data: {
      basename: "page7",
      children: "Content of Page 7",
      colors: "Array(3)",
      content: "Content of Page 7",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Folder overrided site name",
      tags: "Array(3)",
      url: "/pages/new-name/page7/",
    },
    src: {
      asset: false,
      ext: ".page.js",
      path: "/pages/subpage/page7",
      remote: undefined,
    },
  },
  {
    content: "Uint8Array(20)",
    data: {
      basename: "index",
      content: "Uint8Array(20)",
      page: [
        "src",
        "data",
        "asset",
      ],
      url: "/pages/new-name/page7/index.html.br",
    },
    src: {
      asset: true,
      ext: "",
      path: "",
      remote: undefined,
    },
  },
  {
    content: "Content of Page 4 in Overrided site name, from the file /pages/2021-01-02-18-32_page4.page.ts",
    data: {
      basename: "page4",
      children: "Content of Page 4 in Overrided site name, from the file /pages/2021-01-02-18-32_page4.page.ts",
      colors: "Array(3)",
      content: "default",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Overrided site name",
      tags: "Array(1)",
      title: "Page 4",
      url: "/pages/page4/",
    },
    src: {
      asset: false,
      ext: ".page.ts",
      path: "/pages/2021-01-02-18-32_page4",
      remote: undefined,
    },
  },
  {
    content: "Uint8Array(78)",
    data: {
      basename: "index",
      content: "Uint8Array(78)",
      page: [
        "src",
        "data",
        "asset",
      ],
      url: "/pages/page4/index.html.br",
    },
    src: {
      asset: true,
      ext: "",
      path: "",
      remote: undefined,
    },
  },
  {
    content: "<p>Content of Page 6</p>
",
    data: {
      basename: "page6",
      children: "<p>Content of Page 6</p>
",
      colors: "Array(3)",
      content: "Content of Page 6
",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Folder overrided site name",
      tags: "Array(2)",
      title: "Page 6",
      url: "/pages/page6/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/pages/ghost/2021-12-29-page6",
      remote: undefined,
    },
  },
  {
    content: "Uint8Array(29)",
    data: {
      basename: "index",
      content: "Uint8Array(29)",
      page: [
        "src",
        "data",
        "asset",
      ],
      url: "/pages/page6/index.html.br",
    },
    src: {
      asset: true,
      ext: "",
      path: "",
      remote: undefined,
    },
  },
  {
    content: "This is a page",
    data: {
      basename: "static",
      children: "This is a page",
      content: "This is a page",
      date: [],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Default site name",
      tags: "Array(0)",
      url: "/static/",
    },
    src: {
      asset: false,
      ext: ".yml",
      path: "/static",
      remote: undefined,
    },
  },
  {
    content: "Uint8Array(18)",
    data: {
      basename: "index",
      content: "Uint8Array(18)",
      page: [
        "src",
        "data",
        "asset",
      ],
      url: "/static/index.html.br",
    },
    src: {
      asset: true,
      ext: "",
      path: "",
      remote: undefined,
    },
  },
  {
    content: "Uint8Array(52)",
    data: {
      basename: "styles",
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Default site name",
      tags: "Array(0)",
      url: "/styles.css",
    },
    src: {
      asset: true,
      ext: ".css",
      path: "/styles",
      remote: undefined,
    },
  },
  {
    content: "Uint8Array(41)",
    data: {
      basename: "styles",
      content: "Uint8Array(41)",
      page: [
        "src",
        "data",
        "asset",
      ],
      url: "/styles.css.br",
    },
    src: {
      asset: true,
      ext: "",
      path: "",
      remote: undefined,
    },
  },
]
`;

snapshot[`brotli plugin with options 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      isPage: true,
      loader: [AsyncFunction: toml],
    },
    {
      engines: 1,
      ext: ".page.ts",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".page.js",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      engines: 0,
      ext: ".page.json",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".vto",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      ext: ".png",
    },
    {
      ext: ".css",
    },
  ],
  src: [
    "/",
    "/404.md",
    "/_data.yml",
    "/favicon.png",
    "/images",
    "/images/avatar.jpg",
    "/page5.yaml",
    "/pages",
    "/pages/2020-06-21_page2.page.json",
    "/pages/2021-01-02-18-32_page4.page.ts",
    "/pages/_data",
    "/pages/_data.yml",
    "/pages/_data/colors.yml",
    "/pages/_data/documents.ts",
    "/pages/_data/drinks.js",
    "/pages/_data/names.json",
    "/pages/ghost",
    "/pages/ghost/2021-12-29-page6.md",
    "/pages/ghost/_data.yml",
    "/pages/page1.md",
    "/pages/page3.page.js",
    "/pages/subpage",
    "/pages/subpage/_data.yml",
    "/pages/subpage/page7.page.js",
    "/static.yml",
    "/styles.css",
  ],
}
`;

snapshot[`brotli plugin with options 2`] = `
[
  {
    entry: "/favicon.png",
    flags: [],
    outputPath: "/favicon.png",
  },
]
`;

snapshot[`brotli plugin with options 3`] = `
[
  {
    content: "<p>This page is exported to <code>/404.html</code>, not <code>/404/index.html</code></p>
",
    data: {
      basename: "404",
      children: "<p>This page is exported to <code>/404.html</code>, not <code>/404/index.html</code></p>
",
      content: "This page is exported to \`/404.html\`, not \`/404/index.html\`
",
      date: [],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Default site name",
      tags: "Array(0)",
      url: "/404.html",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/404",
      remote: undefined,
    },
  },
  {
    content: "Content of Page 2",
    data: {
      basename: "page2",
      children: "Content of Page 2",
      colors: "Array(3)",
      content: "Content of Page 2",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
        "image",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Folder overrided site name",
      tags: "Array(1)",
      title: "Page 2",
      url: "/overrided-page2/",
    },
    src: {
      asset: false,
      ext: ".page.json",
      path: "/pages/2020-06-21_page2",
      remote: undefined,
    },
  },
  {
    content: "Content of Page 3",
    data: {
      basename: "page3",
      colors: "Array(3)",
      content: "Content of Page 3",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Folder overrided site name",
      tags: "Array(1)",
      title: "Page 3",
      url: "/page_3",
    },
    src: {
      asset: false,
      ext: ".page.js",
      path: "/pages/page3",
      remote: undefined,
    },
  },
  {
    content: "Content of Page 5",
    data: {
      basename: "page5",
      children: "Content of Page 5",
      content: "Content of Page 5",
      date: [],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Default site name",
      tags: "Array(0)",
      title: "Page 5",
      url: "/page5/",
    },
    src: {
      asset: false,
      ext: ".yaml",
      path: "/page5",
      remote: undefined,
    },
  },
  {
    content: "Content of Page 7",
    data: {
      basename: "page7",
      children: "Content of Page 7",
      colors: "Array(3)",
      content: "Content of Page 7",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Folder overrided site name",
      tags: "Array(3)",
      url: "/pages/new-name/page7/",
    },
    src: {
      asset: false,
      ext: ".page.js",
      path: "/pages/subpage/page7",
      remote: undefined,
    },
  },
  {
    content: "Content of Page 4 in Overrided site name, from the file /pages/2021-01-02-18-32_page4.page.ts",
    data: {
      basename: "page4",
      children: "Content of Page 4 in Overrided site name, from the file /pages/2021-01-02-18-32_page4.page.ts",
      colors: "Array(3)",
      content: "default",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Overrided site name",
      tags: "Array(1)",
      title: "Page 4",
      url: "/pages/page4/",
    },
    src: {
      asset: false,
      ext: ".page.ts",
      path: "/pages/2021-01-02-18-32_page4",
      remote: undefined,
    },
  },
  {
    content: "<p>Content of Page 6</p>
",
    data: {
      basename: "page6",
      children: "<p>Content of Page 6</p>
",
      colors: "Array(3)",
      content: "Content of Page 6
",
      date: [],
      documents: "Array(3)",
      drinks: [
        "alcoholic",
        "others",
      ],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      names: "Array(2)",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Folder overrided site name",
      tags: "Array(2)",
      title: "Page 6",
      url: "/pages/page6/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/pages/ghost/2021-12-29-page6",
      remote: undefined,
    },
  },
  {
    content: "This is a page",
    data: {
      basename: "static",
      children: "This is a page",
      content: "This is a page",
      date: [],
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Default site name",
      tags: "Array(0)",
      url: "/static/",
    },
    src: {
      asset: false,
      ext: ".yml",
      path: "/static",
      remote: undefined,
    },
  },
  {
    content: "Uint8Array(52)",
    data: {
      basename: "styles",
      imagick: "Array(1)",
      mergedKeys: [
        "tags",
        "metas",
        "imagick",
      ],
      metas: [
        "title",
        "description",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      site: "Default site name",
      tags: "Array(0)",
      url: "/styles.css",
    },
    src: {
      asset: true,
      ext: ".css",
      path: "/styles",
      remote: undefined,
    },
  },
  {
    content: "Uint8Array(56)",
    data: {
      basename: "styles",
      content: "Uint8Array(56)",
      page: [
        "src",
        "data",
        "asset",
      ],
      url: "/styles.css.br",
    },
    src: {
      asset: true,
      ext: "",
      path: "",
      remote: undefined,
    },
  },
]
`;
